src/super_gradients/training/pre_launch_callbacks/pre_launch_callbacks.py

```diff
@@ -0,0 +1,153 @@
+    this pre_launch_callback will modify cfg.dataset_params.train_dataloader_params.batch_size then pass cfg to
+     Trainer.train_from_config(cfg) and training will continue with the selected batch size.
+
+
+    :param min_batch_size: int, the first batch size to try running forward passes. Should fit memory.
+
+    :param size_step: int, the difference between 2 consecutive batch_size trials.
+
+    :param num_forward_passes: int, number of forward passes (i.e train_loader data iterations inside an epoch).
+     Note that the more forward passes being done, the less the selected batch size is prawn to fail. This is because
+      other then gradients, model computations, data and other fixed gpu memory that is being used- some more gpu memory
+       might be used by the metric objects and PhaseCallbacks.
+
+    :param max_batch_size: int, optional, upper limit of the batch sizes to try. When None, the search will continue until
+     the maximal batch size that does not raise CUDA OUT OF MEMORY is found (deafult=None).
+
+    :param scale_lr: bool, whether to linearly scale cfg.training_hyperparams.initial_lr, i.e multiply by
+     FOUND_BATCH_SIZE/cfg.dataset_params.train_datalaoder_params.batch_size (default=True)
+    """
+
+    def __init__(self, min_batch_size: int, size_step: int, num_forward_passes: int = 3, max_batch_size=None, scale_lr: bool = True):
+        self.scale_lr = scale_lr
+        self.min_batch_size = min_batch_size
+        self.size_step = size_step
+        self.max_batch_size = max_batch_size
+        self.num_forward_passes = num_forward_passes
+
+    def __call__(self, cfg: DictConfig) -> DictConfig:
+
+        # IMPORT IS HERE DUE TO CIRCULAR IMPORT PROBLEM
+        from super_gradients.training.sg_trainer import Trainer
+
+        curr_batch_size = self.min_batch_size
+        # BUILD NETWORK
+        model = models.get(
+            model_name=cfg.architecture,
+            num_classes=cfg.arch_params.num_classes,
+            arch_params=cfg.arch_params,
+            strict_load=cfg.checkpoint_params.strict_load,
+            pretrained_weights=cfg.checkpoint_params.pretrained_weights,
+            checkpoint_path=cfg.checkpoint_params.checkpoint_path,
+            load_backbone=cfg.checkpoint_params.load_backbone,
+        )
+        tmp_cfg = deepcopy(cfg)
+        tmp_cfg.training_hyperparams.batch_accumulate = 1
+        tmp_cfg.training_hyperparams.max_train_batches = self.num_forward_passes
+        tmp_cfg.training_hyperparams.run_validation_freq = 2
+        tmp_cfg.training_hyperparams.silent_mode = True
+        tmp_cfg.training_hyperparams.save_model = False
+        tmp_cfg.training_hyperparams.max_epochs = 1
+        tmp_cfg.training_hyperparams.average_best_models = False
+        tmp_cfg.training_hyperparams.kil_ddp_pgroup_on_end = False
+        tmp_cfg.pre_launch_callbacks_list = []
+
+        while True:
```

Comment:
why not use `for batch_size in range(min_batch_size, min_batch_size*1000, size_step)` ?

Reply:
Why should we limit our users ? Its very reasonable IMO that someone would start with a very small min batch size and large steps that would possibly reach this limit. Also if we replace step_size users still might reach this limit. This is the reason I added the optional max_batch_size.